driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,656 @@
+import static com.mongodb.MongoCredential.OidcRefreshContext;
+import static com.mongodb.MongoCredential.OidcRequestContext;
+import static com.mongodb.MongoCredential.PROVIDER_NAME_KEY;
+import static com.mongodb.MongoCredential.REFRESH_TOKEN_CALLBACK_KEY;
+import static com.mongodb.MongoCredential.REQUEST_TOKEN_CALLBACK_KEY;
+import static com.mongodb.MongoCredential.OidcRefreshCallback;
+import static com.mongodb.MongoCredential.OidcRequestCallback;
+import static com.mongodb.assertions.Assertions.assertFalse;
+import static com.mongodb.assertions.Assertions.assertNotNull;
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.lang.String.format;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public class OidcAuthenticator extends SaslAuthenticator {
+
+    private static final List<String> SUPPORTED_PROVIDERS = Arrays.asList("aws");
+
+    private static final Duration CALLBACK_TIMEOUT = Duration.ofMinutes(5);
+
+    private static final String AWS_WEB_IDENTITY_TOKEN_FILE = "AWS_WEB_IDENTITY_TOKEN_FILE";
+
+    private ServerAddress serverAddress;
+
+    private String connectionLastAccessToken;
+
+    private FallbackState fallbackState = FallbackState.INITIAL;
+
+    private BsonDocument speculativeAuthenticateResponse;
+
+    private Function<byte[], byte[]> evaluateChallengeFunction;
+
+    public OidcAuthenticator(final MongoCredentialWithCache credential,
+            final ClusterConnectionMode clusterConnectionMode, @Nullable final ServerApi serverApi) {
+        super(credential, clusterConnectionMode, serverApi);
+
+        if (getMongoCredential().getAuthenticationMechanism() != MONGODB_OIDC) {
+            throw new MongoException("Incorrect mechanism: " + getMongoCredential().getMechanism());
+        }
+    }
+
+    @Override
+    public String getMechanismName() {
+        return MONGODB_OIDC.getMechanismName();
+    }
+
+    @Override
+    protected SaslClient createSaslClient(final ServerAddress serverAddress) {
+        this.serverAddress = serverAddress;
+        MongoCredentialWithCache mongoCredentialWithCache = getMongoCredentialWithCache();
+        return new OidcSaslClient(mongoCredentialWithCache);
+    }
+
+    @Override
+    @Nullable
+    public BsonDocument createSpeculativeAuthenticateCommand(final InternalConnection connection) {
+        try {
+            OidcRequestCallback requestCallback = getRequestCallback();
+            if (requestCallback == null) {
+                return wrapInSpeculative(prepareAwsTokenFromFile());
+            }
+            String cachedAccessToken = getValidCachedAccessToken();
+            MongoCredentialWithCache mongoCredentialWithCache = getMongoCredentialWithCache();
+            if (cachedAccessToken != null) {
+                connectionLastAccessToken = cachedAccessToken;
+                fallbackState = FallbackState.PHASE_1_CACHED_TOKEN;
```

Comment:
Remove `connectionLastAccessToken`, part of prepareTokenAsJWT. I think fallbackState is reset anyway, so this is left over from when we could tell that spec auth failed.